frontend/src/views/TE/Upload.vue

```diff
@@ -190,6 +209,62 @@ export default class Upload extends Vue {
     }
   }
 
+  getFileHash(file: File) {
+    return new Promise(resolve => {
+      const fileReader = new FileReader();
+      fileReader.onload = function(event: any) {
+        const spark = new Spark.ArrayBuffer();
+        spark.append(event.target.result);
+        resolve(btoa(spark.end(true)));
+      };
+      fileReader.readAsArrayBuffer(file);
+    });
+  }
+
+  async validateFileHash(file: File, task: any) {
+    const fileHash = await this.getFileHash(file);
+    if (!task.versions) return;
+    for (const version of Object.values<any>(task.versions)) {
+      const ref = this.uploadsBucket.ref().child(`/${version.uploadPath}`);
+      const metadata = await ref.getMetadata().catch(e => "error");
+      if (metadata === "error") continue;
+      if (fileHash === metadata.md5Hash) {
+        throw new Error("You had uploaded the same file earlier.");
+      }
+    }
+  }
+
+  async validateUpload(file: File) {
+    validateFlacFile(file);
+    const taskId = getTaskId(file.name);
+    const task = (
+      await firebase
+        .database()
+        .ref(`/TE/tasks/${taskId}`)
+        .once("value")
+    ).val();
+    if (this.currentUser.email !== _.get(task, "assignee.emailAddress")) {
+      throw new Error(`The task ${taskId} is not assigned to you.`);
+    }
+    if (task.status === "Done") {
+      throw new Error(`The task ${taskId} is marked as Done.`);
+    }
+    const lastTaskResolutionTimestamp = _.get(
+      this.getLastVersion(task),
+      "resolution.timestamp"
+    );
+    if (
+      lastTaskResolutionTimestamp &&
+      lastTaskResolutionTimestamp > file.lastModified
+    ) {
+      throw new FileUploadError({
+        props: { taskId },
+        component: "file-time-error"
+      });
+    }
+    await this.validateFileHash(file, task);
```

Comment:
This method is not too big, consider moving its contents here for clarity. I mean, all the validation steps will be directly in this `validateFile` method.